taskcluster/ac_taskgraph/release_promotion.py

```diff
@@ -0,0 +1,172 @@
+                "description": (
+                    "Next version.",
+                ),
+                "default": "",
+            },
+        },
+        "required": ["release_promotion_flavor", "version", "build_number", "next_version"],
+    }
+)
+def release_promotion_action(parameters, graph_config, input, task_group_id, task_id):
+    release_promotion_flavor = input['release_promotion_flavor']
+    promotion_config = graph_config['release-promotion']['flavors'][release_promotion_flavor]
+
+    target_tasks_method = promotion_config['target-tasks-method'].format(
+        project=parameters['project']
+    )
+    rebuild_kinds = input.get('rebuild_kinds') or promotion_config.get('rebuild-kinds', [])
+    do_not_optimize = input.get('do_not_optimize') or promotion_config.get('do-not-optimize', [])
+
+    # make parameters read-write
+    parameters = dict(parameters)
+    # Build previous_graph_ids from ``previous_graph_ids`` or ``revision``.
+    previous_graph_ids = input.get('previous_graph_ids')
+    if not previous_graph_ids:
+        previous_graph_ids = [find_decision_task(parameters, graph_config)]
+
+    # Download parameters from the first decision task
+    parameters = get_artifact(previous_graph_ids[0], "public/parameters.yml")
+    # Download and combine full task graphs from each of the previous_graph_ids.
+    # Sometimes previous relpro action tasks will add tasks, like partials,
+    # that didn't exist in the first full_task_graph, so combining them is
+    # important. The rightmost graph should take precedence in the case of
+    # conflicts.
+    combined_full_task_graph = {}
+    for graph_id in previous_graph_ids:
+        full_task_graph = get_artifact(graph_id, "public/full-task-graph.json")
+        combined_full_task_graph.update(full_task_graph)
+    _, combined_full_task_graph = TaskGraph.from_json(combined_full_task_graph)
+    parameters['existing_tasks'] = find_existing_tasks_from_previous_kinds(
+        combined_full_task_graph, previous_graph_ids, rebuild_kinds
+    )
+    parameters['do_not_optimize'] = do_not_optimize
+    parameters['target_tasks_method'] = target_tasks_method
+    parameters['build_number'] = int(input['build_number'])
+    # When doing staging releases on try, we still want to re-use tasks from
+    # previous graphs.
+    parameters['optimize_target_tasks'] = True
+    parameters['shipping_phase'] = input['release_promotion_flavor']
+
+    version_in_file = read_version_file()
+    parameters['version'] = input['version'] if input.get('version') else read_version_file()
+    version_string = parameters['version']
+    if version_string != version_in_file:
+        raise ValueError("Version given in tag ({}) does not match the one in version.txt ({})".format(version_string, version_in_file))
+    parameters['head_tag'] = 'v{}'.format(version_string)
+
+    parameters['next_version'] = input['next_version']
+
+    version = FenixVersion.parse(version_string)
```

Comment:
We probably also need to change this `FenixVersion` into `MavenVersion`.

Reply:
hah, right. fixed.